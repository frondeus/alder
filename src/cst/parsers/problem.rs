use crate::*;

pub fn recover(parser: impl Parser) -> impl Parser {
    move |state: &mut State| {
        let node = parser.parse(state);
        if !node.is(NodeId::ERROR) {
            state.panic = false;
        }
        node
    }
}

pub fn raise(problem: impl Problem + Clone + 'static, len: usize) -> impl Parser {
    move |state: &mut State| {
        let panic = state.panic;
        let span = state.input.chomp(len);
        match state.last_error() {
            Some(err) if panic => {
                err.span.range.1 += len;
                if let Some(error) = state.errors.last_mut() {
                    error.span.range.1 += len;
                }
                none().parse(state)
            }
            _ if !panic => {
                let problem = Box::new(problem.clone()) as Box<dyn Problem + 'static>;
                let context = state
                    .nodes
                    .iter()
                    .flat_map(|node| node.all_names_with_span())
                    .filter(|(name, _)| !NodeId::NO_CONTEXT.contains(name))
                    .map(|(name, span)| ParseErrorContext::new(name, span))
                    .collect();
                state
                    .errors
                    .push(ParseError::new(problem, span.clone(), context));
                state.panic = true;
                Node::error(span)
            }
            _ => Node::error(span),
        }
    }
}
