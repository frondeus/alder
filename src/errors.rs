use log::error;
use std::error::Error as StdError;

struct SourceIter<'a> {
    current: Option<&'a (dyn StdError + 'static)>,
}
impl<'a> Iterator for SourceIter<'a> {
    type Item = &'a (dyn StdError + 'static);
    fn next(&mut self) -> Option<Self::Item> {
        let current = self.current;
        self.current = self.current.and_then(StdError::source);
        current
    }
}

pub fn source_iter(error: &impl StdError) -> impl Iterator<Item = &(dyn StdError + 'static)> {
    SourceIter {
        current: error.source(),
    }
}

pub trait ResultExt<T> {
    fn unwrap_display(self) -> T;
    fn unwrap_display_err(self) -> String;
    fn unwrap_log(self) -> Option<T>;
}

impl<T, E> ResultExt<T> for Result<T, E>
where
    E: StdError,
{
    fn unwrap_display(self) -> T {
        match self {
            Err(err) => {
                let mut cause = format!("{}", err);
                for source in source_iter(&err) {
                    cause += &format!("\ncaused by: {}", source);
                }
                panic!("\n{}", cause)
            }
            Ok(o) => o,
        }
    }

    fn unwrap_display_err(self) -> String {
        if let Err(err) = self {
            let mut cause = format!("{}", err);
            for source in source_iter(&err) {
                cause += &format!("\ncaused by: {}", source);
            }
            cause
        } else {
            panic!("Expected error")
        }
    }

    fn unwrap_log(self) -> Option<T> {
        match self {
            Err(err) => {
                error!("{}", err);
                for source in source_iter(&err) {
                    error!("caused by: {}", source);
                }
                None
            }
            Ok(o) => Some(o),
        }
    }
}
