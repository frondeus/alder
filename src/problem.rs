use crate::offset::Offset;
use crate::{Parser, Result};
use colored::{ColoredString, Colorize};
use std::fmt::Error;
use std::fmt::Formatter;
use std::fmt::{Debug, Display};
use std::iter::repeat;
use std::ops::{Deref, DerefMut};

pub struct Ctx<P1, C> {
    pub(crate) c: C,
    pub(crate) p: P1,
}
impl<'a, P, C, P1> Parser<'a, C, P> for Ctx<P1, C>
where
    P1: Parser<'a, C, P>,
    C: Copy,
{
    type T = P1::T;

    fn parse(&self, i: &'a str) -> Result<'a, Self::T, C, P> {
        self.p.parse(i).map_err(|mut e| {
            e.iter_mut().for_each(|e| e.add_context(self.c, i));
            e
        })
    }
}

pub fn context<'a, P1, P, C, T>(c: C, p: P1) -> impl Parser<'a, C, P, T = T>
where
    P1: Parser<'a, C, P, T = T>,
    C: Copy,
{
    Ctx { c, p }
}

#[derive(Debug)]
pub struct DeadEnds<'a, C, P> {
    pub ends: Vec<DeadEnd<'a, C, P>>,
}

impl<'a, C, P> DeadEnds<'a, C, P> {
    pub fn failure(&self) -> bool {
        self.ends.iter().any(|e| e.failure)
    }
}

impl<'a, P, C> Deref for DeadEnds<'a, C, P> {
    type Target = Vec<DeadEnd<'a, C, P>>;

    fn deref(&self) -> &Self::Target {
        &self.ends
    }
}

impl<'a, P, C> DerefMut for DeadEnds<'a, C, P> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.ends
    }
}

impl<'a, P, C> From<Vec<DeadEnd<'a, C, P>>> for DeadEnds<'a, C, P> {
    fn from(ends: Vec<DeadEnd<'a, C, P>>) -> Self {
        Self { ends }
    }
}

#[derive(Debug)]
pub struct DeadEnd<'a, C, P> {
    pub problem: P,
    pub failure: bool,
    pub input: &'a str,
    pub context: Vec<(&'a str, C)>,
}

impl<'a, P, C> DeadEnd<'a, C, P> {
    pub fn new(problem: P, input: &'a str) -> Self {
        Self {
            problem,
            input,
            failure: false,
            context: vec![],
        }
    }

    pub fn vec(problem: P, input: &'a str) -> DeadEnds<'a, C, P> {
        DeadEnds {
            ends: vec![Self::new(problem, input)],
        }
    }

    pub fn add_context(&mut self, c: C, input: &'a str) {
        //if !self.failure {
        //self.context.insert(0, (input, c));
        self.context.push((input, c));
        //}
    }

    pub fn cut(&mut self) {
        self.failure = true;
    }
}

pub fn cut<'a, C, P, T>(p: impl Parser<'a, C, P, T = T>) -> impl Parser<'a, C, P, T = T> {
    move |i| {
        p.parse(i).map_err(|mut d| {
            d.iter_mut().for_each(|i| i.cut());
            d
        })
    }
}

pub struct DisplayInputEntry<'a> {
    whre: &'a str,
    desc: Option<ColoredString>,
}

pub struct DisplayInput<'a> {
    input: &'a str,
    entries: Vec<DisplayInputEntry<'a>>,
}

impl<'a> DisplayInput<'a> {
    pub fn new(input: &'a str, whre: &'a str) -> Self {
        Self {
            input,
            entries: vec![DisplayInputEntry { whre, desc: None }],
        }
    }
    pub fn with_desc(mut self, desc: impl Into<ColoredString>) -> Self {
        self.entries[0].desc = Some(desc.into()); //TODO
        self
    }

    fn eol(line: usize, lines: &[&'a str]) -> &'static str {
        if lines.len() <= line + 1 {
            "\\EOF"
        } else {
            "\\n"
        }
    }

    /// Output: (line, column)
    fn position(&self, lines: &[&'a str], whre: &'a str) -> (usize, usize) {
        let mut offset = self.input.offset(whre);
        let mut line = 0;
        let mut column = 0;

        for (j, l) in lines.iter().enumerate() {
            if offset <= l.len() {
                line = j;
                column = offset;
                break;
            } else {
                offset = offset - l.len() - 1;
                line = j + 1;
            }
        }

        (line, column)
    }
}

impl<'a> Display for DisplayInput<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::result::Result<(), Error> {
        let input = self.input;
        let lines = input.lines().collect::<Vec<_>>();

        for entry in self.entries.iter() {
            let (line, column) = self.position(&lines, entry.whre);

            if line > 0 {
                if let Some(prev) = lines.get(line - 1) {
                    write!(f, "{}", format!("{}  | ", line - 1).cyan())?;
                    write!(f, "{}", (prev).bright_white())?;
                    writeln!(f, "{}", "\\n".bright_black())?;
                }
            }

            write!(f, "{}", format!("{}  | ", line).cyan())?;
            if let Some(this) = lines.get(line) {
                write!(f, "{}", this.bright_white())?;
            }
            writeln!(f, "{}", Self::eol(line, &lines).bright_black())?;
            write!(f, "{}", "   | ".cyan())?;
            if column > 0 {
                write!(f, "{}", &repeat(' ').take(column).collect::<String>())?;
            }
            let desc = entry.desc.clone().unwrap_or_else(|| "^".normal());
            writeln!(f, "{}", &desc.bold())?;
            if let Some(next) = lines.get(line + 1) {
                write!(f, "{}", format!("{}  | ", line + 1).cyan())?;
                write!(f, "{}", (next).bright_white())?;
                writeln!(f, "{}", Self::eol(line + 1, &lines).bright_black())?;
            }
        }

        Ok(())
    }
}
