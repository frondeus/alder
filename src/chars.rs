use crate::problem::DeadEnd;
use crate::Parser;

pub fn take<'a, C, P: Clone>(len: usize, problem: P) -> impl Parser<'a, C, P, T = &'a str> {
    move |i: &'a str| {
        let i_len = i.len();
        if i_len < len {
            return Err(DeadEnd::vec(problem.clone(), i));
        }
        Ok((&i[0..len], &i[len..]))
    }
}

pub fn char_where<'a, C, P: Clone>(
    f: impl Fn(char) -> bool,
    problem: P,
) -> impl Parser<'a, C, P, T = char> {
    let parser = take(1, problem.clone()).map(|s| s.chars().next());
    move |i| {
        let (first, rest) = parser.parse(i)?;
        match first {
            Some(letter) if f(letter) => Ok((letter, rest)),
            _ => Err(DeadEnd::vec(problem.clone(), i)),
        }
    }
}

pub fn chomp_while<'a, C, P>(f: impl Fn(char) -> bool) -> impl Parser<'a, C, P, T = &'a str> {
    move |i: &'a str| {
        let mut len = 0;
        loop {
            let c = &i[len..].chars().next();
            match c {
                Some(letter) if f(*letter) => {
                    len += 1;
                }
                _ => break,
            }
        }

        let result = &i[0..len];
        let rest = &i[len..];
        Ok((result, rest))
    }
}

pub fn token<'a, C, P: Clone>(expected: char, problem: P) -> impl Parser<'a, C, P, T = char> {
    let parser = take(1, problem.clone()).map(|s| s.chars().next());
    move |i| {
        let (first, rest) = parser.parse(i)?;
        match first {
            Some(letter) if letter == expected => Ok((letter, rest)),
            _ => Err(DeadEnd::vec(problem.clone(), i)),
        }
    }
}

pub fn tag<C, P: Clone>(expected: &str, problem: P) -> impl Parser<C, P, T = &str> {
    move |i| {
        let len = expected.len();
        let (o, r) = take(len, problem.clone()).parse(i)?;
        if o == expected {
            return Ok((o, r));
        }
        Err(DeadEnd::vec(problem.clone(), i))
    }
}

pub fn spaces<'a, C, P>() -> impl Parser<'a, C, P, T = &'a str> {
    chomp_while(|f| f.is_whitespace())
}
/*
impl<'a, C> Parser<'a, C, CharProblem<'a>> for char {
    type T = char;

    fn parse(&self, i: &'a str) -> Result<'a, Self::T, C, CharProblem<'a>> {
        token(*self).parse(i)
    }
}


impl<'a, C> Parser<'a, C, CharProblem<'a>> for &'a str {
    type T = &'a str;

    fn parse(&self, i: &'a str) -> Result<'a, Self::T, C, CharProblem<'a>> {
        tag(self).parse(i)
    }
}
*/
