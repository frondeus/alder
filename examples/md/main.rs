#![feature(drain_filter)]
use crate::report::Report;
use alder::chars::{char_where, chomp_while, spaces, token};
use alder::errors::ResultExt;
use alder::problem::cut;
use alder::{delimited, opt, or, preceded, recognize, terminated, IntoParser, Parser};
use colored::Colorize;
use derive_more::Display;
use std::collections::BTreeMap;
use std::fmt::Debug;
use test_case::test_case;

mod report;

#[derive(Debug)]
enum Value<'a> {
    ConstNum(f64),
    ConstStr(&'a str),
    ConstArr(Vec<Value<'a>>),
    ConstObj(BTreeMap<&'a str, Value<'a>>),
}

#[derive(Debug, Display, Clone, PartialEq)]
pub enum Problem {
    #[display(fmt = "I expected number like 0, 1, -2, 0.5")]
    ExpectedLiteralNumber,

    #[display(fmt = "I expected '-'")]
    ExpectedMinus,

    #[display(fmt = "I expected '.'")]
    ExpectedDot,

    #[display(fmt = r#"I expected '"'"#)]
    ExpectedDoublequote,

    #[display(fmt = r#"I expected '['"#)]
    ExpectedOpenBracket,

    #[display(fmt = r#"I expected ']'"#)]
    ExpectedCloseBracket,

    #[display(fmt = r#"I expected ','"#)]
    ExpectedComma,

    #[display(fmt = "I expected '{{'")]
    ExpectedCurlyOpenBracket,

    #[display(fmt = "I expected '}}'")]
    ExpectedCurlyCloseBracket,

    #[display(fmt = "I expected '='")]
    ExpectedEqual,

    #[display(fmt = "I expected identifier")]
    ExpectedIdentifier,
}

impl Problem {
    pub fn name(&self) -> &'static str {
        use Problem::*;
        match self {
            ExpectedLiteralNumber => "num",
            ExpectedIdentifier => "ident",
            ExpectedMinus => "-",
            ExpectedDot => ".",
            ExpectedDoublequote => r#"""#,
            ExpectedOpenBracket => "[",
            ExpectedCloseBracket => "]",
            ExpectedComma => ",",
            ExpectedCurlyOpenBracket => "{",
            ExpectedCurlyCloseBracket => "}",
            ExpectedEqual => "=",
        }
    }
}

#[derive(Debug, Clone, Copy, Display)]
pub enum Context {
    //#[display(fmt = "I tried to parse number {}", r#""<name> = <value>".white().bold()"#)]
    #[display(fmt = "I tried to parse const value")]
    ConstValue,

    #[display(fmt = "I tried to parse const number")]
    ConstNumber,

    #[display(fmt = "I tried to parse const string")]
    ConstStr,

    #[display(fmt = "I tried to parse const array")]
    ConstArray,

    #[display(fmt = "I tried to parse const obj")]
    ConstObj,
}

fn c_value<'a>() -> impl Parser<'a, Context, Problem, T = Value<'a>> {
    |i| {
        preceded(
            spaces(),
            (c_number(10), c_string(), c_array(), c_obj()).or(),
        )
        .ctx(Context::ConstValue)
        .parse(i)
    }
}

fn c_string<'a>() -> impl Parser<'a, Context, Problem, T = Value<'a>> {
    preceded(
        token('"', Problem::ExpectedDoublequote),
        cut(terminated(
            chomp_while(|c| c != '"'),
            token('"', Problem::ExpectedDoublequote),
        )),
    )
    .map(Value::ConstStr)
    .ctx(Context::ConstStr)
}

fn c_obj<'a>() -> impl Parser<'a, Context, Problem, T = Value<'a>> {
    preceded(
        token('{', Problem::ExpectedCurlyOpenBracket),
        cut(terminated(
            delimited(c_obj_pair(), token(',', Problem::ExpectedComma)),
            token('}', Problem::ExpectedCurlyCloseBracket),
        )
        .map(Value::ConstObj)),
    )
    .ctx(Context::ConstObj)
}

fn c_obj_pair<'a>() -> impl Parser<'a, Context, Problem, T = (&'a str, Value<'a>)> {
    (
        recognize((
            char_where(
                |c| c != '=' && !c.is_whitespace(),
                Problem::ExpectedIdentifier,
            ),
            chomp_while(|c| c != '=' && !c.is_whitespace()),
        )),
        spaces(),
        token('=', Problem::ExpectedEqual),
        c_value(),
    )
        .map(|(a, _, _, b)| (a, b))
}

fn c_array<'a>() -> impl Parser<'a, Context, Problem, T = Value<'a>> {
    preceded(
        token('[', Problem::ExpectedOpenBracket),
        cut(terminated(
            delimited(c_value(), token(',', Problem::ExpectedComma)),
            token(']', Problem::ExpectedCloseBracket),
        )
        .map(Value::ConstArr)),
    )
    .ctx(Context::ConstArray)
}

fn c_number<'a>(radix: u32) -> impl Parser<'a, Context, Problem, T = Value<'a>> {
    recognize((
        opt(token('-', Problem::ExpectedMinus)),
        char_where(move |c| c.is_digit(radix), Problem::ExpectedLiteralNumber),
        chomp_while(move |c| c.is_digit(radix)),
        opt((
            token('.', Problem::ExpectedDot),
            chomp_while(move |c| c.is_digit(radix)),
        )),
    ))
    .map(|number| {
        let n = number.parse::<f64>().unwrap_or_default();
        Value::ConstNum(n)
    })
    .ctx(Context::ConstNumber)
}

#[test_case(c_value(),   "--5"    ; "minus minus digit" )]
#[test_case(c_value(),   "a"    ; "non" )]
#[test_case(c_value(),   r#""a"#    ; "string missing end" )]
#[test_case(c_value(),   r#"[5, 6"#    ; "array missing end" )]
#[test_case(c_value(),   r#"[5, 6,"#    ; "array missing end trailing comma" )]
#[test_case(c_value(),   r#"{"#    ; "obj missing end" )]
#[test_case(c_value(),   r#"{foo = 5"#    ; "obj missing end with pair" )]
#[test_case(c_value(),   r#"{foo = 5,"#    ; "obj missing end with pair trailing comma" )]
fn test_errors<'a, T: Debug>(p: impl Parser<'a, Context, Problem, T = T>, input: &'a str) {
    let res = p.parse(input).map_err(|ends| Report::new(input, ends));

    dbg!(&res);
    let err = res.unwrap_display_err();

    insta::assert_display_snapshot!(err);
}

#[test_case(c_value(),   "-5" => matches (Value::ConstNum(-5.0), _)        ; "minus digit" )]
#[test_case(c_value(),   "5"  => matches (Value::ConstNum(5.0), _)         ; "digit" )]
#[test_case(c_value(),   "5.4" => matches (Value::ConstNum(5.4), _)      ; "double" )]
#[test_case(c_value(),   r#""foo""# => matches (Value::ConstStr("foo"), _)      ; "string" )]
#[test_case(c_value(),   r#"[5, 6]"# => matches (Value::ConstArr(_), _)    ; "array" )]
#[test_case(c_value(),   r#"[5, 6,]"# => matches (Value::ConstArr(_), _)    ; "array trailing comma" )]
#[test_case(c_value(),   r#"[]"# => matches (Value::ConstArr(_), _)    ; "empty array" )]
#[test_case(c_value(),   r#"{}"# => matches (Value::ConstObj(_), _)    ; "empty obj" )]
#[test_case(c_value(),   r#"{foo = 5}"# => matches (Value::ConstObj(_), _)    ; "obj" )]
#[test_case(c_value(),   r#"{foo = 5,}"# => matches (Value::ConstObj(_), _)    ; "obj trailing comma" )]
fn test<'a, T: Debug>(p: impl Parser<'a, Context, Problem, T = T>, input: &'a str) -> (T, &'a str) {
    p.parse(input)
        .map_err(|ends| Report::new(input, ends))
        .unwrap_display()
}

fn main() {}
