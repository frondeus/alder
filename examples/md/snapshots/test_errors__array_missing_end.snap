---
source: examples/md/main.rs
expression: err
---
[1;31m-- SYNTAX ERROR --[0m
I tried to parse const array
[36m0  | [0m[97m[5, 6[0m[90m\EOF[0m
[36m   | [0m[1;36m^ started here[0m
[36m0  | [0m[97m[5, 6[0m[90m\EOF[0m
[36m   | [0m     [1m[31m^ unexpected token.[0m[1m I expected ']'[0m


